CEP XXXX: Conditional dependencies, extras and flags (#111)
conda/ceps: cep-XXX1.md

The lack of this feature in conda packaging has forced packagers to (inefficiently) translate this expressivity into a set of different package variants which only differ in their dependency metadata. For example, the following runtime requirements of a pure Python package:

```toml
# ...
requires_python = ">=3.8"
requirements = [
  "typing-extensions; python_version < '3.9'",
  "pywin32; sys_platform == 'win32'"
  "requests",
]
```

Can only be converted into conda packages by making a potentially `noarch: python` package (because it's pure Python) a non-noarch package distributed as as many as 25 platforms (five supported Python versions, times five platforms). This way, the conditional dependency is resolved (in a way) at build time by rendering the dependencies into the granular artifacts. For example, the `typing-extensions` dependency would only be present in the variants built for Python 3.8, and we would only find `pywin32` in the packages published to the `win-*` subdirs.

> The adoption of conditional dependencies in conda would allow packagers to express the same complexity in a single `noarch: python` package, because these conditions would be evaluated at "solve time". Providing these capabilities requires new `MatchSpec` syntax and a bump in the repodata version to allow the new syntax in the `depends` field.

Reviewer: I kind of like the idea of not breaking the schema completely, but base updates on additive fields. In that case, why don't we let go of the `"packages.whatever": {key: package record}` idea and devise a schema under a new key that is better future-proofed? For example a new top-level key that is `v3` or similar: its value can be whatever we need it to be.

What we need to nail down is how to transition from version to version without leaving users behind, and without having partial availability. Some sort of "migrated_at"  timestamp so the client errors can be informative?

> The adoption of conditional dependencies in conda would allow packagers to express the same complexity in a single `noarch: python` package, because these conditions would be evaluated at "solve time". Providing these capabilities requires new `MatchSpec` syntax and a bump in the repodata version to allow the new syntax in the `depends` field.

Reviewer 2: The problem is that that doesnt really allow structural changes to json; you already have to parse the depends to know how to parse the depends.

> The adoption of conditional dependencies in conda would allow packagers to express the same complexity in a single `noarch: python` package, because these conditions would be evaluated at "solve time". Providing these capabilities requires new `MatchSpec` syntax and a bump in the repodata version to allow the new syntax in the `depends` field.

Author: There is actually a `__conda` virtual package IIRC - which obviously doesn't work / translate well to other clients. Changing that for a "conda-spec" virtual package would make sense IMO.

However, as Bas pointed out, I don't think that would work for this transition. Good material for another CEP though! :)

> The adoption of conditional dependencies in conda would allow packagers to express the same complexity in a single `noarch: python` package, because these conditions would be evaluated at "solve time". Providing these capabilities requires new `MatchSpec` syntax and a bump in the repodata version to allow the new syntax in the `depends` field.

Reviewer: There you go: https://github.com/conda/ceps/pull/146

# CEP XXXX - Conditional dependencies

<table>
<tr><td> Title </td><td> Conditional dependencies </td>
<tr><td> Status </td><td> Draft </td></tr>
<tr><td> Author(s) </td><td>
  Wolf Vollprecht &lt;w.vollprecht@gmail.com&gt;,
  Bas Zalmstra &lt;bas@prefix.dev&gt;,
  Jaime Rodríguez-Guerra &lt;jaime.rogue@gmail.com&gt;
</td></tr>
<tr><td> Created </td><td> Feb 5, 2025</td></tr>
<tr><td> Updated </td><td> Mar 9, 2026</td></tr>
<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/111 </td></tr>
<tr><td> Implementation </td><td> TBD </td></tr>
<tr><td> Requires </td><td> https://github.com/conda/ceps/pull/146 </td></tr>
</table>

## Abstract

This CEP proposes the introduction of conditional dependencies in the conda ecosystem. The work comprises two changes: the addition of a new keyword argument, `when`, to the `MatchSpec` syntax, and its evaluation at solve time. The value of the `when` argument is always a logical expression of one or more MatchSpec queries targeting the solver context.

## Motivation

> Conditional dependencies are often requested in the conda ecosystem. After all, an equivalent feature exist in Python packaging in the form of "environment markers" (introduced in [PEP 508](https://peps.python.org/pep-0508/)), so it's a reasonable expectation. There are even traces of planned support of it in `conda.models.match_spec` (see commit [`ab33436`](https://github.com/conda/conda/pull/7606/changes/ab33436591cd2f8ce9c792a50a412d1d02c75504)) in the context of ["pip interoperability"](https://github.com/conda/conda/issues/7053).

Reviewer 3: ```suggestion
Conditional dependencies are often requested in the conda ecosystem. After all, an equivalent feature exists in Python packaging in the form of "environment markers" (introduced in [PEP 508](https://peps.python.org/pep-0508/)), so it's a reasonable expectation. There are even traces of planned support of it in `conda.models.match_spec` (see commit [`ab33436`](https://github.com/conda/conda/pull/7606/changes/ab33436591cd2f8ce9c792a50a412d1d02c75504)) in the context of ["pip interoperability"](https://github.com/conda/conda/issues/7053).
```

Classic `MatchSpec` queries have traditionally matched fields present in candidate package records. For example, `python>=3.10` would match records with package name `python` and version greater or equal than `3.10`. The `when` field is a bit different in that sense, since it doesn't target a `when` field in the target records. Instead, its value is a condition that will match a context comprised of (potentially) other records. This is a deviation for the `MatchSpec` design but we still feel that the ergonomics offered by this syntax are worth the extra complexity.

## Specification

This CEP extends [CEP 29](./cep-0029.md) with a new keyword, `when`.

### Syntax

A conditional dependency is defined as a `MatchSpec` string that features a `when` keyword, the value of which MUST be a string that encodes a logical expression of one or more `MatchSpec` queries. The logical expression follows a Python-like syntax: `MatchSpec` strings MAY be joined with operators `and` (logical AND) and `or` (logical OR), and grouped within parentheses `()` for precedence overrides.

> The inner `MatchSpec` queries inside `when` MUST be expressed in their square brackets syntax, with the exception of simple `name` and `version` queries that MAY be expressed as `{name}{operator}{version}` (no space separators). These inner `MatchSpec` queries MUST NOT feature their own `when` field.

Reviewer 4: We may need to be more restrictive on what types of `MatchSpec` clauses can appear in `when`, especially since square bracket clauses as described in CEP 29 are relatively unconstrained and could lead to highly ambiguous parse trees. E.g., you could have square bracket clauses that themselves use quotes, parentheses, `and`, `or`, etc.

For the purposes of this CEP, it might be easier if we just start by allowing only `{name}{operator}{version}` clauses and expand the set of allowable expressions in follow up CEPs as the need arises.

Reviewer 2: We have a parser that works. From what perspective do you want to restrict this more? We tried to avoid ambiguity by restricting to `{name}{operator}{version}` already.

You probably have a few ideas of what ambiguous cases look like? Maybe you can share those?

Reviewer: Taking one last look here, does `{name}{operator}{version}` include `python>=3,<4` forms or only `python>=3`?

Reviewer 5: Re-raised on https://github.com/conda/ceps/pull/164#discussion_r3146270057 since this CEP was split out and voting has been called.

Reviewer 5: > We have a parser that works.

@baszalmstra Hey Bas, that seems uncalled for given Cheng's valid feedback about the CEP. It reads a little passive-aggressive to me, and I'd appreciate it if we followed up on this thread in #164. Given this circumstance of your comment, calling for a vote without resolving it seems a little unfortunate.

> If necessary, the `when` value MUST be quoted to avoid parsing ambiguities.

Reviewer 4: I would drop the "If necessary, " part and just say "The `when` value MUST be quoted to avoid parsing ambiguities."

Reviewer 2: ```suggestion
The `when` value MUST be quoted to avoid parsing ambiguities.
```

```

In an input file:

```yaml
name: env-with-conditions
channels: [conda-forge]
dependencies:
- python
- numpy>=2[when="python>=3.10"]
```

```toml
[dependencies]
python = "*"
numpy = { version=">=2", when="python>=3.10" }
```

In a recipe file:

```yaml:
# ...
requirements:
  run:
    - python
    - numpy>=2[when="python>=3.10"]
```

## Backwards compatibility

Adding a new field to the `MatchSpec` keywords is a backwards incompatible change. For example, in `conda`:

```shell
$ conda create -d "python[when=__unix]"
InvalidMatchSpec: Invalid spec 'python[when=python]': Invalid spec 'python[when=python]': Cannot match on field(s): {'when'}
```

When used as part of the runtime requirements of a newly built package, the conditional dependency must be included in the `depends` field of the resulting repodata record. Since this new syntax is backwards incompatible with older clients, the resulting `repodata.json` documents (or a derivative, like the sharded forms) MUST NOT include records of packages whose `info/index.json` features a value of `schema_version` equals to or greater than 3.

## Rejected ideas

> - `...; if syntax` and other backwards compatible proposal that would be inadequate because then the condition would be ignored, hence creating invalid solutions.

Reviewer 3: Do we mean
```suggestion
- `...; if ...` syntax, and other backwards compatible proposals, are rejected because the condition would be ignored by solvers which do not implement this CEP, resulting in invalid solutions.
```
